File: rc4_decryptor.c
#include "rc4.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define DEFAULT_KEY "MySecretKey2026"
#define ENCRYPTED_FILE "file.txt"
#define DECRYPTED_FILE "file_decrypted.txt"
#define BACKUP_FILE "file.txt.backup"

void print_hex(const char *label, const uint8_t *data, size_t len, size_t max_print) {
    printf("%s", label);
    size_t print_len = (len < max_print) ? len : max_print;
    for (size_t i = 0; i < print_len; i++) {
        printf("%02x", data[i]);
    }
    if (len > max_print) {
        printf("... (%zu more bytes)", len - max_print);
    }
    printf("\n");
}

int verify_files(const char *file1, const char *file2) {
    FILE *fp1 = fopen(file1, "rb");
    FILE *fp2 = fopen(file2, "rb");
    
    if (!fp1 || !fp2) {
        if (fp1) fclose(fp1);
        if (fp2) fclose(fp2);
        return 0;
    }

    int result = 1;
    uint8_t buf1[4096];
    uint8_t buf2[4096];
    size_t read1, read2;

    while (1) {
        read1 = fread(buf1, 1, sizeof(buf1), fp1);
        read2 = fread(buf2, 1, sizeof(buf2), fp2);
        
        if (read1 != read2) {
            result = 0;
            break;
        }
        if (read1 == 0) {
            break;  /* Both reached EOF */
        }
        if (memcmp(buf1, buf2, read1) != 0) {
            result = 0;
            break;
        }
    }

    fclose(fp1);
    fclose(fp2);
    return result;
}

int main(int argc, char *argv[]) {
    const char *key_str = (argc > 1) ? argv[1] : DEFAULT_KEY;
    const char *encrypted_file = ENCRYPTED_FILE;
    const char *decrypted_file = DECRYPTED_FILE;
    FILE *fp;
    uint8_t *buffer = NULL;
    size_t file_size = 0;
    size_t bytes_read = 0;
    long pos;

    printf("==================================================\n");
    printf("RC4 File Decryptor (C Implementation)\n");
    printf("==================================================\n");
    printf("Encrypted file:  %s\n", encrypted_file);
    printf("Decrypted file:  %s\n", decrypted_file);
    printf("Key:             %s\n", key_str);
    printf("--------------------------------------------------\n");

    /* Open encrypted file */
    fp = fopen(encrypted_file, "rb");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open '%s' for reading.\n", encrypted_file);
        fprintf(stderr, "Please run the encryptor first to create an encrypted file.\n");
        return 1;
    }

    /* Get file size */
    fseek(fp, 0, SEEK_END);
    pos = ftell(fp);
    if (pos < 0) {
        fprintf(stderr, "Error: Cannot determine file size.\n");
        fclose(fp);
        return 1;
    }
    file_size = (size_t)pos;
    fseek(fp, 0, SEEK_SET);

    if (file_size == 0) {
        fprintf(stderr, "Error: File is empty.\n");
        fclose(fp);
        return 1;
    }

    /* Allocate buffer */
    buffer = (uint8_t *)malloc(file_size);
    if (!buffer) {
        fprintf(stderr, "Error: Memory allocation failed for %zu bytes.\n", file_size);
        fclose(fp);
        return 1;
    }

    /* Read encrypted data */
    bytes_read = fread(buffer, 1, file_size, fp);
    fclose(fp);

    if (bytes_read != file_size) {
        fprintf(stderr, "Error: Failed to read complete file (read %zu of %zu bytes).\n",
                bytes_read, file_size);
        free(buffer);
        return 1;
    }

    printf("Encrypted file size: %zu bytes\n", file_size);
    print_hex("Encrypted content preview: ", buffer, file_size, 100);

    /* Decrypt the data (RC4 is symmetric) */
    rc4((const uint8_t *)key_str, strlen(key_str), buffer, file_size);

    /* Write decrypted data to output file */
    fp = fopen(decrypted_file, "wb");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open '%s' for writing.\n", decrypted_file);
        free(buffer);
        return 1;
    }

    size_t bytes_written = fwrite(buffer, 1, file_size, fp);
    fclose(fp);

    if (bytes_written != file_size) {
        fprintf(stderr, "Error: Failed to write complete file (wrote %zu of %zu bytes).\n",
                bytes_written, file_size);
        free(buffer);
        return 1;
    }

    free(buffer);

    printf("--------------------------------------------------\n");
    printf("Decryption completed successfully!\n");
    printf("Decrypted file size: %zu bytes\n", file_size);

    /* Show decrypted preview */
    fp = fopen(decrypted_file, "rb");
    if (fp) {
        buffer = (uint8_t *)malloc(file_size);
        if (buffer) {
            fread(buffer, 1, file_size, fp);
            fclose(fp);
            print_hex("Decrypted content preview: ", buffer, file_size, 100);
            free(buffer);
        }
    }

    /* Verify against original backup if it exists */
    printf("--------------------------------------------------\n");
    if (verify_files(BACKUP_FILE, decrypted_file)) {
        printf("VERIFICATION PASSED: Decrypted file matches original exactly!\n");
    } else {
        printf("VERIFICATION FAILED: Decrypted file does NOT match original!\n");
        printf("Note: Create a backup before encryption: cp file.txt %s\n", BACKUP_FILE);
        return 1;
    }

    return 0;
}

File: rc4_decryptor.py
#!/usr/bin/env python3
"""
RC4 File Decryptor
Decrypts the contents of file.txt using RC4 stream cipher.
"""
import sys
import os
from typing import Union
from rc4 import rc4


def decrypt_file(input_file: str, output_file: str, key: Union[bytes, str]) -> bool:
    """
    Decrypt a file using RC4.
    
    Args:
        input_file: Path to the encrypted input file
        output_file: Path to the output file (can be same as input for in-place)
        key: Decryption key (bytes or string) - must match encryption key
    
    Returns:
        True if successful, False otherwise
    """
    try:
        # Read the encrypted file in binary mode
        with open(input_file, 'rb') as f:
            encrypted_data = f.read()
        
        print(f"Read {len(encrypted_data)} encrypted bytes from {input_file}")
        
        # Convert key to bytes if string
        if isinstance(key, str):
            key = key.encode()
        
        # Decrypt the data (RC4 is symmetric)
        decrypted_data = rc4(key, encrypted_data)
        
        # Write decrypted data to output file
        with open(output_file, 'wb') as f:
            f.write(decrypted_data)
        
        print(f"Written {len(decrypted_data)} decrypted bytes to {output_file}")
        return True
    
    except FileNotFoundError:
        print(f"Error: File '{input_file}' not found.")
        return False
    except PermissionError:
        print(f"Error: Permission denied accessing '{input_file}' or '{output_file}'.")
        return False
    except Exception as e:
        print(f"Error during decryption: {e}")
        return False


def verify_files(file1: str, file2: str) -> bool:
    """
    Verify two files have identical content.
    
    Args:
        file1: Path to first file
        file2: Path to second file
    
    Returns:
        True if files are identical, False otherwise
    """
    try:
        with open(file1, 'rb') as f1, open(file2, 'rb') as f2:
            data1 = f1.read()
            data2 = f2.read()
        
        return data1 == data2
    except Exception as e:
        print(f"Error during verification: {e}")
        return False


def main():
    # Configuration
    ENCRYPTED_FILE = "file.txt"      # The encrypted file
    DECRYPTED_FILE = "file_decrypted.txt"  # Output decrypted file
    KEY = "MySecretKey2026"          # Must match encryption key
    ORIGINAL_FILE = "file.txt.backup"  # Backup of original for verification (optional)
    
    print("=" * 50)
    print("RC4 File Decryptor")
    print("=" * 50)
    print(f"Encrypted file:  {ENCRYPTED_FILE}")
    print(f"Decrypted file:  {DECRYPTED_FILE}")
    print(f"Key:             {KEY}")
    print("-" * 50)
    
    # Check if encrypted file exists
    if not os.path.exists(ENCRYPTED_FILE):
        print(f"Error: {ENCRYPTED_FILE} does not exist.")
        print("Please run the encryptor first to create an encrypted file.")
        sys.exit(1)
    
    # Show encrypted content preview
    with open(ENCRYPTED_FILE, 'rb') as f:
        encrypted = f.read()
    
    print(f"Encrypted file size: {len(encrypted)} bytes")
    if len(encrypted) > 0:
        preview = encrypted[:100]
        print(f"Encrypted content preview: {preview!r}")
        if len(encrypted) > 100:
            print("  ... (truncated)")
    
    # Decrypt the file
    success = decrypt_file(ENCRYPTED_FILE, DECRYPTED_FILE, KEY)
    
    if success:
        print("-" * 50)
        print("Decryption completed successfully!")
        
        # Show decrypted content preview
        with open(DECRYPTED_FILE, 'rb') as f:
            decrypted = f.read()
        
        print(f"Decrypted file size: {len(decrypted)} bytes")
        if len(decrypted) > 0:
            preview = decrypted[:100]
            print(f"Decrypted content preview: {preview!r}")
            if len(decrypted) > 100:
                print("  ... (truncated)")
        
        # Verify against original if backup exists
        if os.path.exists(ORIGINAL_FILE):
            print("-" * 50)
            print("Verifying against original backup...")
            if verify_files(ORIGINAL_FILE, DECRYPTED_FILE):
                print("✓ VERIFICATION PASSED: Decrypted file matches original exactly!")
            else:
                print("✗ VERIFICATION FAILED: Decrypted file does NOT match original!")
                sys.exit(1)
        else:
            print("-" * 50)
            print("Note: No original backup found for verification.")
            print(f"Create a backup before encryption: cp file.txt {ORIGINAL_FILE}")
    
    else:
        print("-" * 50)
        print("Decryption failed!")
        sys.exit(1)


if __name__ == "__main__":
    # Allow key to be passed as command line argument
    if len(sys.argv) > 1:
        KEY = sys.argv[1]
        print(f"Using key from command line: {KEY}")
    
    main()